File: ExpressControls/AppRadioButton.vb
' Follow steps 1a or 1b and then 2 to use this custom control in a XAML file.
'
' Step 1a) Using this custom control in a XAML file that exists in the current project.
' Add this XmlNamespace attribute to the root element of the markup file where it is 
' to be used:
'
'     xmlns:MyNamespace="clr-namespace:ExpressControls"
'
'
' Step 1b) Using this custom control in a XAML file that exists in a different project.
' Add this XmlNamespace attribute to the root element of the markup file where it is 
' to be used:
'
'     xmlns:MyNamespace="clr-namespace:ExpressControls;assembly=ExpressControls"
'
' You will also need to add a project reference from the project where the XAML file lives
' to this project and Rebuild to avoid compilation errors:
'
'     Right click on the target project in the Solution Explorer and
'     "Add Reference"->"Projects"->[Browse to and select this project]
'
'
' Step 2)
' Go ahead and use your control in the XAML file. Note that Intellisense in the
' XML editor does not currently work on custom controls and its child elements.
'
'     <MyNamespace:AppRadioButton/>
'

Imports System.Windows.Controls.Primitives


Public Class AppRadioButton
    Inherits RadioButton

    Shared Sub New()
        'This OverrideMetadata call tells the system that this element wants to provide a style that is different than its base class.
        'This style is defined in themes\generic.xaml
        DefaultStyleKeyProperty.OverrideMetadata(GetType(AppRadioButton), New FrameworkPropertyMetadata(GetType(AppRadioButton)))
    End Sub

    Public Shared ReadOnly CornerRadiusProperty As DependencyProperty = DependencyProperty.Register(NameOf(CornerRadius), GetType(CornerRadius), GetType(AppRadioButton), New PropertyMetadata(Nothing))

    Public Property CornerRadius As CornerRadius
        Get
            Return CType(GetValue(CornerRadiusProperty), CornerRadius)
        End Get
        Set(ByVal value As CornerRadius)
            SetValue(CornerRadiusProperty, value)
        End Set
    End Property

    Public Shared ReadOnly GapMarginProperty As DependencyProperty = DependencyProperty.Register(NameOf(GapMargin), GetType(Thickness), GetType(AppRadioButton), New PropertyMetadata(Nothing))

    Public Property GapMargin As Thickness
        Get
            Return CType(GetValue(GapMarginProperty), Thickness)
        End Get
        Set(ByVal value As Thickness)
            SetValue(GapMarginProperty, value)
        End Set
    End Property

    Public Shared ReadOnly IconSizeProperty As DependencyProperty = DependencyProperty.Register(NameOf(IconSize), GetType(Double), GetType(AppRadioButton), New PropertyMetadata(Nothing))

    Public Property IconSize As Double
        Get
            Return CType(GetValue(IconSizeProperty), Double)
        End Get
        Set(ByVal value As Double)
            SetValue(IconSizeProperty, value)
        End Set
    End Property

End Class

File: ExpressControls/AppRadioButton.cs
using System;
using System.Windows;
using System.Windows.Controls;

namespace ExpressControls
{
    /// <summary>
    /// Follow steps 1a or 1b and then 2 to use this custom control in a XAML file.
    ///
    /// Step 1a) Using this custom control in a XAML file that exists in the current project.
    /// Add this XmlNamespace attribute to the root element of the markup file where it is
    /// to be used:
    ///
    ///     xmlns:MyNamespace="clr-namespace:ExpressControls"
    ///
    ///
    /// Step 1b) Using this custom control in a XAML file that exists in a different project.
    /// Add this XmlNamespace attribute to the root element of the markup file where it is
    /// to be used:
    ///
    ///     xmlns:MyNamespace="clr-namespace:ExpressControls;assembly=ExpressControls"
    ///
    /// You will also need to add a project reference from the project where the XAML file lives
    /// to this project and Rebuild to avoid compilation errors:
    ///
    ///     Right click on the target project in the Solution Explorer and
    ///     "Add Reference"->"Projects"->[Browse to and select this project]
    ///
    ///
    /// Step 2)
    /// Go ahead and use your control in the XAML file.
    ///
    ///     <MyNamespace:AppRadioButton/>
    ///
    /// </summary>
    public class AppRadioButton : RadioButton
    {
        static AppRadioButton()
        {
            DefaultStyleKeyProperty.OverrideMetadata(
                typeof(AppRadioButton),
                new FrameworkPropertyMetadata(typeof(AppRadioButton))
            );
        }

        public static readonly DependencyProperty CornerRadiusProperty =
            DependencyProperty.Register(
                nameof(CornerRadius),
                typeof(CornerRadius),
                typeof(AppRadioButton),
                new PropertyMetadata(null)
            );

        public CornerRadius CornerRadius
        {
            get { return (CornerRadius)GetValue(CornerRadiusProperty); }
            set { SetValue(CornerRadiusProperty, value); }
        }

        public static readonly DependencyProperty GapMarginProperty = DependencyProperty.Register(
            nameof(GapMargin),
            typeof(Thickness),
            typeof(AppRadioButton),
            new PropertyMetadata(null)
        );

        public Thickness GapMargin
        {
            get { return (Thickness)GetValue(GapMarginProperty); }
            set { SetValue(GapMarginProperty, value); }
        }

        public static readonly DependencyProperty IconSizeProperty = DependencyProperty.Register(
            nameof(IconSize),
            typeof(double),
            typeof(AppRadioButton),
            new PropertyMetadata(null)
        );

        public double IconSize
        {
            get { return Convert.ToDouble(GetValue(IconSizeProperty)); }
            set { SetValue(IconSizeProperty, value); }
        }
    }
}
